### ISSUE TITLE
error: aesm_service returned error: 31

### ISSUE THREAD

**Issue by alexander-lindner** [intel/linux-sgx#775]
Hello everyone,

when I start an enclave with RA, it fails with this message (the code runs on an old v1 notebook):
```
error: aesm_service returned error: 31
error: load_enclave() failed with error -1
```
I guess, that the code 31 corresponds with this enum: https://github.com/intel/linux-sgx/blob/master/psw/ae/inc/internal/aesm_error.h#L66. However, I don't know what the stmt `AESM_UNRECOGNIZED_PLATFORM` is suppose to tell me as I don't find more information's about that? Can you give me an hint ? Maybe, for future, this error handling could be added with more informations?

Some more debug information
```
systemctl status aesmd

Jan 14 08:17:04 sgxserver aesm_service[411095]: [ADMIN]EPID Provisioning initiated
Jan 14 08:17:04 sgxserver aesm_service[411095]: The Request ID is fbb0ebf75da744cf82670994534d3345
Jan 14 08:17:05 sgxserver aesm_service[411095]: The Request ID is 009e97cb152d44d7a67c2746f1c2b323
Jan 14 08:17:05 sgxserver aesm_service[411095]: [ADMIN]EPID Provisioning protocol error reported by Backend (6)
Jan 14 08:17:05 sgxserver aesm_service[411095]: [ADMIN]EPID Provisioning failed


is-sgx-available 
SGX supported by CPU: true
SGX1 (ECREATE, EENTER, ...): true
SGX2 (EAUG, EACCEPT, EMODPR, ...): true
Flexible Launch Control (IA32_SGXPUBKEYHASH{0..3} MSRs): true
SGX extensions for virtualizers (EINCVIRTCHILD, EDECVIRTCHILD, ESETCONTEXT): false
Extensions for concurrent memory management (ETRACKC, ELDBC, ELDUC, ERDINFO): false
CET enclave attributes support (See Table 37-5 in the SDM): false
Key separation and sharing (KSS) support (CONFIGID, CONFIGSVN, ISVEXTPRODID, ISVFAMILYID report fields): true
Max enclave size (32-bit): 0x80000000
Max enclave size (64-bit): 0x100000000000000
EPC size: 0x1fcf800000
SGX driver loaded: true
AESMD installed: true
SGX PSW/libsgx installed: true
```

**Comment by llly** [comment]
The production SGX EPID backend server reports that you platform can not be attested.
It's not a production platform, cannot pass Remote attestation.

**Comment by alexander-lindner** [comment]
I am not sure how I should interpret your response `It's not a production platform`. Can you specify it? Is it possible to switch to an development EPID backend server?

**Comment by ScottR-Intel** [comment]
Hi Alexander.

Based on the size of EPC and the fact that SGX2/EDMM and KSS is supported, it seems you're running on a 3rd Gen Xeon Scalable processor, aka Ice Lake Server (ICX).  ICX and all future Xeon Scalables do not/will not support EPID based provisioning/RA.  You will need to use DCAP/ECDSA based RA.  More info at the links below:

https://github.com/intel/SGXDataCenterAttestationPrimitives
https://www.intel.com/content/www/us/en/developer/articles/technical/quote-verification-attestation-with-intel-sgx-dcap.html
https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/

Scott

**Comment by alexander-lindner** [comment]
Hi Scott,

you are right, it's a new server with bought last December with an Icelake based processor: 
`Intel(R) Xeon(R) Gold 5318S CPU ` (https://ark.intel.com/content/www/us/en/ark/products/215279/intel-xeon-gold-5318s-processor-36m-cache-2-10-ghz.html?wapkw=5318S).

So i guess your answer is right. I will install dcap and test it. Thanks alot.

However, is there any test/command/c-function I can run on an unknown server to check if EPID is supported. Also, I didn't find any materials for this CPU regarding EPID support. How do I know when buying a new one that EPID is supported?

Alex

**Comment by ScottR-Intel** [comment]
Hi again Alex.

Please see this recent SGX Forum answer:  https://community.intel.com/t5/Intel-Software-Guard-Extensions/whether-attestation-services-are-available-on-my-processor/m-p/1352875/highlight/true#M5113

Let me know if you have more questions regarding this.

Regards.

Scott

**Comment by alexander-lindner** [comment]
Hi Scott,

thanks for the answer. I managed to install DCAP and now I can RA my enclave and verify it, so my problem is fixed. 

> Please see this recent SGX Forum answer: https://community.intel.com/t5/Intel-Software-Guard-Extensions/whether-attestation-services-are-available-on-my-processor/m-p/1352875/highlight/true#M5113

So there is no easy way which can check for EPID/DCAP support. Therefore, one always needs to do some research. I guess I have to live with that...

Regards
Alex
